src/scan-iterator.ts

```diff
@@ -138,18 +33,106 @@ export class ScanResult<K> implements AsyncIterable<JSONValue> {
   }
 
   values(): AsyncIterableIterator<JSONValue> {
-    return this._newIterator('value');
+    return this._newIterator(VALUE);
   }
 
   keys(): AsyncIterableIterator<K> {
-    return this._newIterator('key');
+    return this._newIterator(KEY);
   }
 
   entries(): AsyncIterableIterator<[K, JSONValue]> {
-    return this._newIterator('entry');
+    return this._newIterator(ENTRY);
   }
 
   private _newIterator<V>(kind: ScanIterableKind): AsyncIterableIterator<V> {
-    return new ScanIterator<V>(kind, ...this._args);
+    return scanIterator(kind, ...this._args);
   }
 }
+
+async function* scanIterator<V>(
+  kind: ScanIterableKind,
+  options: ScanOptions | undefined,
+  invoke: Invoke,
+  getTransaction: () => Promise<IdCloser> | IdCloser,
+  shouldCloseTranscation: boolean,
+): AsyncGenerator<V> {
+  const transaction = await getTransaction();
+  throwIfClosed(transaction);
+
+  let controller!: ReadableStreamDefaultController<V>;
+  const stream = new ReadableStream({
+    start: c => {
+      controller = c;
+    },
+  });
+  const reader: ReadableStreamDefaultReader<V> = stream.getReader();
+
+  // No await. We want the loading to happen in the background. load
+  // communicates with this function using controller.
+  load(kind, options, transaction.id, controller, invoke);
+
+  try {
+    while (true) {
```

Comment:
How come `scan_raw` is all synchronous? Why doesn't it have to wait for IDB?

If it is all synchronous I don't see reason to use a callback either. Let's just return a synchronous iterator then.

Reply:
The initial read is async, but because we don't have chunking yet (no prolly trees), the initial read ends up getting all the data. We haven't made the internal functions async yet just because we haven't needed to.

We will eventually have prolly trees and make this async, but I don't think that addresses the issue of wanting data to load lazily, but *also* needing to be able to close the transaction before completing the scan.

I'm sure there's some way to invalidate the iterator if the transaction closes before it's done, but it seems complex due to ownership inside Rust. The open iteration will have a read lock on the transaction, and I am assuming that closing the transaction will require a write lock. This means that naively the closeTransaction RPC would hang until the iterator releases its read lock.

So you'd have to wire something up inside the JS to get the iterator to release its read lock when the tx function returns. Which ... at that point, I'm just like forget it, let's just have the user JS provide a callback 😅.

Let me know if you think I've got any of this wrong.

Reply:
For-of loops calls `return` on the iterator to close it. When that happens the Rust iterator can also stop. But I do agree that getting this to work with Rust seems hard.